File: agentic_pipeline/python_originals/validate.py
"""
Validation utilities for the lesson authoring pipeline.

Validates JSON artifacts against schemas and cross-checks references.
Uses jsonschema if available, falls back to structural checks.
"""

import json
from pathlib import Path

SCHEMA_DIR = Path(__file__).parent / "schemas"


def _load_schema(name):
    with open(SCHEMA_DIR / f"{name}.json") as f:
        return json.load(f)


def _try_jsonschema(data, schema):
    """Attempt jsonschema validation. Returns errors list or None if unavailable."""
    try:
        import jsonschema
        v = jsonschema.Draft202012Validator(schema)
        return [e.message for e in v.iter_errors(data)]
    except ImportError:
        return None


def validate_plan(plan):
    """
    Validate a lesson plan dict against structural requirements.

    Checks (in order):
      1. JSON schema validation if jsonschema is installed.
      2. Presence of "meta" with "title".
      3. Presence of "problem" with "text".
      4. Presence of "nodes" list.
      5. Uniqueness of node IDs (including branch act IDs).
      6. All gate "after_act" values reference a known act ID.
      7. All marker "after_act" values (if set) reference a known act ID.

    @param plan: dict — the lesson plan to validate
    @returns: list[str] — error messages; empty list means the plan is valid.
              Returns on first schema error batch rather than accumulating with
              structural errors, to avoid redundant reporting.
    @effects: no mutation of plan
    """
    errors = []

    schema = _load_schema("lesson_plan")
    schema_errors = _try_jsonschema(plan, schema)
    if schema_errors is not None:
        errors.extend(schema_errors)
        if errors:
            return errors

    # Structural checks
    if "meta" not in plan:
        errors.append("Missing 'meta' field")
    elif "title" not in plan.get("meta", {}):
        errors.append("Missing 'meta.title'")

    if "problem" not in plan:
        errors.append("Missing 'problem' field")
    elif "text" not in plan.get("problem", {}):
        errors.append("Missing 'problem.text'")

    if "nodes" not in plan:
        errors.append("Missing 'nodes' field")
        return errors

    # Check node IDs are unique
    ids = set()
    act_ids = set()
    for node in plan["nodes"]:
        nid = node.get("id")
        ntype = node.get("type")
        if nid:
            if nid in ids:
                errors.append(f"Duplicate node ID: {nid}")
            ids.add(nid)
        if ntype == "act":
            act_ids.add(nid)
            # Collect branch act IDs too
            for wp in node.get("wrong_path_acts", []):
                if isinstance(wp, dict):
                    act_ids.add(wp.get("id"))

    # Check gate after_act references
    for node in plan["nodes"]:
        if node.get("type") == "gate":
            after = node.get("after_act")
            if after and after not in act_ids:
                errors.append(f"Gate {node.get('id')} references non-existent act: {after}")

    # Check marker after_act references
    for node in plan["nodes"]:
        if node.get("type") == "marker":
            after = node.get("after_act")
            if after and after not in act_ids:
                errors.append(f"Marker '{node.get('label')}' references non-existent act: {after}")

    return errors


def validate_act_spec(act_spec, plan=None, plan_node=None):
    """
    Validate an act spec dict against structural and cross-reference requirements.

    Checks:
      1. JSON schema validation if jsonschema is installed.
      2. Presence of "act_id".
      3. Presence and non-emptiness of "beats".
      4. Each beat has a "say" field.
      5. If plan is provided and plan["viz_requirements"]["actions"] is non-empty,
         every viz_action method in every beat must appear in the plan's declared
         action set.

    @param act_spec: dict — the act spec to validate
    @param plan: dict|None — the lesson plan, used for viz action cross-checking.
                 If None, cross-checking is skipped.
    @returns: list[str] — error messages; empty list means the spec is valid.
    @effects: no mutation of act_spec or plan
    """
    errors = []

    schema = _load_schema("act_spec")
    schema_errors = _try_jsonschema(act_spec, schema)
    if schema_errors is not None:
        errors.extend(schema_errors)
        if errors:
            return errors

    if "act_id" not in act_spec:
        errors.append("Missing 'act_id'")
    if "beats" not in act_spec:
        errors.append("Missing 'beats'")
        return errors
    if not act_spec["beats"]:
        errors.append("Act has no beats")

    for i, beat in enumerate(act_spec["beats"]):
        if "say" not in beat:
            errors.append(f"Beat {i} missing 'say' field")
        elif not isinstance(beat["say"], str) or not beat["say"].strip():
            errors.append(f"Beat {i} 'say' must be non-empty string")

    # Duplicate beat detection — catches the act_worker producing repeated narration.
    seen = {}
    for i, beat in enumerate(act_spec.get("beats", [])):
        say = beat.get("say", "")
        if say and say in seen:
            errors.append(f"Beat {i} has duplicate 'say' text of beat {seen[say]} (act_worker must produce unique narration)")
        else:
            seen[say] = i

    # Cross-check viz actions against plan if available
    if plan and plan.get("viz_requirements", {}).get("actions"):
        known = {a["method"] for a in plan["viz_requirements"]["actions"]}
        for i, beat in enumerate(act_spec.get("beats", [])):
            for va in beat.get("viz_actions", []):
                if va.get("method") and va["method"] not in known:
                    errors.append(f"Beat {i} uses unknown viz action: {va['method']}")

    # Plan-node-specific checks: beat count + per-beat viz_action presence.
    # The planner's beat_outline is the structural source of truth — the act
    # worker must produce one beat per outline entry, no more, no fewer.
    if plan_node is not None:
        outline = plan_node.get("beat_outline", []) or []
        beats = act_spec.get("beats", [])
        if outline and len(beats) != len(outline):
            errors.append(
                f"Beat count mismatch: produced {len(beats)}, plan outline has "
                f"{len(outline)}. Act worker must emit exactly one beat per outline entry."
            )
        # If the outline declares viz_actions for a beat, the spec MUST include them.
        for i, out in enumerate(outline):
            required_methods = out.get("viz_actions") or []
            if not required_methods or i >= len(beats):
                continue
            beat = beats[i]
            got_methods = [va.get("method") for va in beat.get("viz_actions", []) if va.get("method")]
            missing = [m for m in required_methods if m not in got_methods]
            if missing:
                errors.append(
                    f"Beat {i} missing required viz_actions from plan: {missing}. "
                    f"Got: {got_methods or '[]'}."
                )

    return errors


def validate_gate_spec(gate_spec, plan=None):
    """
    Validate a gate spec dict against structural requirements.

    Checks:
      1. JSON schema validation if jsonschema is installed.
      2. Presence of "gate_type".
      3. Presence of "after_act".
      4. Type-specific required fields:
           quiz:    "question", "options", "correct" (correct=0 is valid)
           fill-in: "prompt", "blank"

    NOTE: "gate_id" is intentionally NOT validated here — it is injected by the
    orchestrator from the lesson plan node's "id" field and never comes from the LLM.
    Checking for it here would validate pipeline-injected data, not LLM output.

    @param gate_spec: dict — the gate spec to validate
    @param plan: dict|None — reserved for future cross-reference checks; currently unused.
    @returns: list[str] — error messages; empty list means the spec is valid.
    @effects: no mutation of gate_spec or plan
    """
    errors = []

    schema = _load_schema("gate_spec")
    schema_errors = _try_jsonschema(gate_spec, schema)
    if schema_errors is not None:
        errors.extend(schema_errors)
        if errors:
            return errors

    if "gate_type" not in gate_spec:
        errors.append("Missing 'gate_type'")
    if "after_act" not in gate_spec:
        errors.append("Missing 'after_act'")
    # gate_id is pipeline-injected — never check it here

    gt = gate_spec.get("gate_type")
    if gt == "quiz":
        if not gate_spec.get("question"):
            errors.append("Quiz gate missing 'question'")
        if not gate_spec.get("options"):
            errors.append("Quiz gate missing 'options'")
        if gate_spec.get("correct") is None:
            errors.append("Quiz gate missing 'correct'")
    elif gt == "fill-in":
        prompt = gate_spec.get("prompt")
        if not prompt:
            errors.append("Fill-in gate missing 'prompt'")
        elif "[___]" not in prompt:
            errors.append(
                "Fill-in gate 'prompt' missing [___] blank marker — the engine "
                "splits on [___] to render the input field; without it no input "
                "is drawn and the student cannot answer."
            )
        if not gate_spec.get("blank"):
            errors.append("Fill-in gate missing 'blank'")

    return errors


def validate_viz_spec(viz_spec, plan=None, all_acts=None):
    """
    Validate a viz spec dict against structural and cross-reference requirements.

    Checks:
      1. JSON schema validation if jsonschema is installed.
      2. Presence of "mode".
      3. Mode-specific required fields:
           "custom_code":    "code" must be present and non-empty
           "mobject_plugin": "mobject_plugin_code" must be present and non-empty
           "preset":         "preset" must be present and non-empty
      4. If all_acts is provided and viz_spec["actions_implemented"] is present,
         every viz_action method used in any beat of any act must appear in
         actions_implemented.

    @param viz_spec: dict — the viz spec to validate
    @param plan: dict|None — reserved; currently unused.
    @param all_acts: dict|None — mapping act_id → act_spec dict, used for
                     cross-checking viz action method names. If None, skipped.
    @returns: list[str] — error messages; empty list means the spec is valid.
    @effects: no mutation of viz_spec, plan, or all_acts
    """
    errors = []

    schema = _load_schema("viz_spec")
    schema_errors = _try_jsonschema(viz_spec, schema)
    if schema_errors is not None:
        errors.extend(schema_errors)
        if errors:
            return errors

    mode = viz_spec.get("mode")
    if not mode:
        errors.append("Missing 'mode'")
        return errors

    if mode == "custom_code" and not viz_spec.get("code"):
        errors.append("custom_code mode requires 'code' field")
    if mode == "mobject_plugin" and not viz_spec.get("mobject_plugin_code"):
        errors.append("mobject_plugin mode requires 'mobject_plugin_code' field")
    if mode == "preset" and not viz_spec.get("preset"):
        errors.append("preset mode requires 'preset' field")

    # Cross-check: all viz actions used in acts should be implemented
    if all_acts and viz_spec.get("actions_implemented"):
        implemented = set(viz_spec["actions_implemented"])
        for act_id, act in all_acts.items():
            for beat in act.get("beats", []):
                for va in beat.get("viz_actions", []):
                    method = va.get("method")
                    if method and method not in implemented:
                        errors.append(f"Act {act_id} uses viz action '{method}' not in actions_implemented")

    return errors


def validate_all(plan, act_specs, gate_specs, viz_spec):
    """
    Run all validators on a complete set of pipeline artifacts.

    @param plan: dict — parsed lesson_plan.json
    @param act_specs: dict mapping act_id (str) → act_spec dict
    @param gate_specs: dict mapping gate_id (str) → gate_spec dict
    @param viz_spec: dict|None — parsed viz_spec.json, or None to skip viz validation
    @returns: dict mapping category str → list[str] of error messages.
              Categories are: "plan", "act:<act_id>", "gate:<gate_id>", "viz".
              Only categories with at least one error appear in the returned dict.
              Empty dict means all artifacts are valid.
    @effects: no mutation of any input dict
    """
    results = {}

    plan_errors = validate_plan(plan)
    if plan_errors:
        results["plan"] = plan_errors

    for act_id, spec in act_specs.items():
        errs = validate_act_spec(spec, plan)
        if errs:
            results[f"act:{act_id}"] = errs

    for gate_id, spec in gate_specs.items():
        errs = validate_gate_spec(spec, plan)
        if errs:
            results[f"gate:{gate_id}"] = errs

    if viz_spec:
        errs = validate_viz_spec(viz_spec, plan, act_specs)
        if errs:
            results["viz"] = errs

    return results


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="Validate pipeline JSON artifacts")
    parser.add_argument("--plan", help="Path to lesson_plan.json")
    parser.add_argument("--act", action="append", help="Path to an act_spec.json (repeatable)")
    parser.add_argument("--gate", action="append", help="Path to a gate_spec.json (repeatable)")
    parser.add_argument("--viz", help="Path to viz_spec.json")
    args = parser.parse_args()

    plan = None
    if args.plan:
        with open(args.plan) as f:
            plan = json.load(f)
        errors = validate_plan(plan)
        if errors:
            print(f"PLAN errors:")
            for e in errors:
                print(f"  - {e}")
        else:
            print(f"PLAN: OK")

    for path in (args.act or []):
        with open(path) as f:
            spec = json.load(f)
        errors = validate_act_spec(spec, plan)
        label = spec.get("act_id", path)
        if errors:
            print(f"ACT {label} errors:")
            for e in errors:
                print(f"  - {e}")
        else:
            print(f"ACT {label}: OK")

    for path in (args.gate or []):
        with open(path) as f:
            spec = json.load(f)
        errors = validate_gate_spec(spec, plan)
        label = spec.get("gate_id", path)
        if errors:
            print(f"GATE {label} errors:")
            for e in errors:
                print(f"  - {e}")
        else:
            print(f"GATE {label}: OK")

    if args.viz:
        with open(args.viz) as f:
            spec = json.load(f)
        errors = validate_viz_spec(spec, plan)
        if errors:
            print(f"VIZ errors:")
            for e in errors:
                print(f"  - {e}")
        else:
            print(f"VIZ: OK")

File: video-old-code/agentic_pipeline/validate.ts
// LAYER 1: validation (TypeScript port of validate.py).
//
// Validates the pipeline's JSON artifacts (plan, act specs, gate specs, viz spec)
// against structural rules and cross references. each validator returns a list of
// error strings, empty means valid.
//
// NOTE: validate.py optionally uses the Python `jsonschema` package for full JSON
// Schema validation, and falls back to the structural checks below when it isn't
// installed. this port has no jsonschema wired up, so tryJsonschema always reports
// "unavailable" and we go straight to the structural checks. that matches Python's
// behavior when jsonschema isn't installed. the structural checks are a faithful port.
//
// Inputs are typed `any` on purpose: these functions exist to check possibly
// malformed data, same as the Python dicts they came from.

import * as fs from "fs";
import * as path from "path";
import { pathToFileURL } from "url";

const SCHEMA_DIR = path.join(import.meta.dirname, "schemas");

function loadSchema(name: string): any {
  return JSON.parse(fs.readFileSync(path.join(SCHEMA_DIR, `${name}.json`), "utf8"));
}

// jsonschema isn't available here, so this always returns null (unavailable), which
// sends every validator down the structural-check path. same as Python's ImportError.
function tryJsonschema(_data: any, _schema: any): string[] | null {
  return null;
}


// Validate a lesson plan: required fields, unique node ids, and that gate/marker
// after_act values point at real acts.
export function validatePlan(plan: any): string[] {
  const errors: string[] = [];

  const schema = loadSchema("lesson_plan");
  const schemaErrors = tryJsonschema(plan, schema);
  if (schemaErrors !== null) {
    errors.push(...schemaErrors);
    if (errors.length) return errors;
  }

  // structural checks
  if (!("meta" in plan)) errors.push("Missing 'meta' field");
  else if (!("title" in (plan.meta || {}))) errors.push("Missing 'meta.title'");

  if (!("problem" in plan)) errors.push("Missing 'problem' field");
  else if (!("text" in (plan.problem || {}))) errors.push("Missing 'problem.text'");

  if (!("nodes" in plan)) {
    errors.push("Missing 'nodes' field");
    return errors;
  }

  // node ids must be unique (branch act ids count too)
  const ids = new Set<any>();
  const actIds = new Set<any>();
  for (const node of plan.nodes) {
    const nid = node.id;
    const ntype = node.type;
    if (nid) {
      if (ids.has(nid)) errors.push(`Duplicate node ID: ${nid}`);
      ids.add(nid);
    }
    if (ntype === "act") {
      actIds.add(nid);
      for (const wp of node.wrong_path_acts || []) {
        if (wp && typeof wp === "object") actIds.add(wp.id);
      }
    }
  }

  // gate after_act references
  for (const node of plan.nodes) {
    if (node.type === "gate") {
      const after = node.after_act;
      if (after && !actIds.has(after)) {
        errors.push(`Gate ${node.id} references non-existent act: ${after}`);
      }
    }
  }

  // marker after_act references
  for (const node of plan.nodes) {
    if (node.type === "marker") {
      const after = node.after_act;
      if (after && !actIds.has(after)) {
        errors.push(`Marker '${node.label}' references non-existent act: ${after}`);
      }
    }
  }

  return errors;
}


// Validate one act spec: act_id, non-empty beats, each beat has a say, no duplicate
// narration, and (optionally) viz actions match the plan and the plan node outline.
export function validateActSpec(actSpec: any, plan: any = null, planNode: any = null): string[] {
  const errors: string[] = [];

  const schema = loadSchema("act_spec");
  const schemaErrors = tryJsonschema(actSpec, schema);
  if (schemaErrors !== null) {
    errors.push(...schemaErrors);
    if (errors.length) return errors;
  }

  if (!("act_id" in actSpec)) errors.push("Missing 'act_id'");
  if (!("beats" in actSpec)) {
    errors.push("Missing 'beats'");
    return errors;
  }
  if (!actSpec.beats || actSpec.beats.length === 0) errors.push("Act has no beats");

  (actSpec.beats as any[]).forEach((beat, i) => {
    if (!("say" in beat)) errors.push(`Beat ${i} missing 'say' field`);
    else if (typeof beat.say !== "string" || !beat.say.trim()) {
      errors.push(`Beat ${i} 'say' must be non-empty string`);
    }
  });

  // duplicate beat detection: catches the act_worker repeating narration.
  // null-proto map so prototype-name says ("toString"/"valueOf"/"__proto__"/...)
  // don't false-match against inherited members, matching the Python dict.
  const seen: Record<string, number> = Object.create(null);
  (actSpec.beats || []).forEach((beat: any, i: number) => {
    const say = beat.say ?? "";
    if (say && say in seen) {
      errors.push(
        `Beat ${i} has duplicate 'say' text of beat ${seen[say]} (act_worker must produce unique narration)`
      );
    } else {
      seen[say] = i;
    }
  });

  // cross-check viz actions against the plan if we have one
  if (plan && plan.viz_requirements?.actions?.length) {
    const known = new Set((plan.viz_requirements.actions as any[]).map((a) => a.method));
    (actSpec.beats || []).forEach((beat: any, i: number) => {
      for (const va of beat.viz_actions || []) {
        if (va.method && !known.has(va.method)) {
          errors.push(`Beat ${i} uses unknown viz action: ${va.method}`);
        }
      }
    });
  }

  // plan-node-specific checks: beat count + per-beat viz_action presence.
  // the planner outline is the structural source of truth, one beat per entry.
  if (planNode !== null) {
    const outline = planNode.beat_outline || [];
    const beats = actSpec.beats || [];
    if (outline.length && beats.length !== outline.length) {
      errors.push(
        `Beat count mismatch: produced ${beats.length}, plan outline has ${outline.length}. ` +
          "Act worker must emit exactly one beat per outline entry."
      );
    }
    outline.forEach((out: any, i: number) => {
      const requiredMethods = out.viz_actions || [];
      if (!requiredMethods.length || i >= beats.length) return;
      const beat = beats[i];
      const gotMethods = (beat.viz_actions || [])
        .map((va: any) => va.method)
        .filter((m: any) => m);
      const missing = requiredMethods.filter((m: any) => !gotMethods.includes(m));
      if (missing.length) {
        errors.push(
          `Beat ${i} missing required viz_actions from plan: ${missing}. ` +
            `Got: ${gotMethods.length ? gotMethods : "[]"}.`
        );
      }
    });
  }

  return errors;
}


// Validate one gate spec: gate_type, after_act, and the type-specific required fields.
// gate_id is intentionally NOT checked here (the pipeline injects it, not the LLM).
export function validateGateSpec(gateSpec: any, _plan: any = null): string[] {
  const errors: string[] = [];

  const schema = loadSchema("gate_spec");
  const schemaErrors = tryJsonschema(gateSpec, schema);
  if (schemaErrors !== null) {
    errors.push(...schemaErrors);
    if (errors.length) return errors;
  }

  if (!("gate_type" in gateSpec)) errors.push("Missing 'gate_type'");
  if (!("after_act" in gateSpec)) errors.push("Missing 'after_act'");
  // gate_id is pipeline-injected, never check it here

  const gt = gateSpec.gate_type;
  if (gt === "quiz") {
    if (!gateSpec.question) errors.push("Quiz gate missing 'question'");
    if (!gateSpec.options) errors.push("Quiz gate missing 'options'");
    if (gateSpec.correct === undefined || gateSpec.correct === null) {
      errors.push("Quiz gate missing 'correct'");
    }
  } else if (gt === "fill-in") {
    const prompt = gateSpec.prompt;
    if (!prompt) {
      errors.push("Fill-in gate missing 'prompt'");
    } else if (!prompt.includes("[___]")) {
      errors.push(
        "Fill-in gate 'prompt' missing [___] blank marker — the engine " +
          "splits on [___] to render the input field; without it no input " +
          "is drawn and the student cannot answer."
      );
    }
    if (!gateSpec.blank) errors.push("Fill-in gate missing 'blank'");
  }

  return errors;
}


// Validate one viz spec: mode plus the mode-specific required field, and (optionally)
// that every viz action used across acts is listed in actions_implemented.
export function validateVizSpec(vizSpec: any, _plan: any = null, allActs: any = null): string[] {
  const errors: string[] = [];

  const schema = loadSchema("viz_spec");
  const schemaErrors = tryJsonschema(vizSpec, schema);
  if (schemaErrors !== null) {
    errors.push(...schemaErrors);
    if (errors.length) return errors;
  }

  const mode = vizSpec.mode;
  if (!mode) {
    errors.push("Missing 'mode'");
    return errors;
  }

  if (mode === "custom_code" && !vizSpec.code) errors.push("custom_code mode requires 'code' field");
  if (mode === "mobject_plugin" && !vizSpec.mobject_plugin_code) {
    errors.push("mobject_plugin mode requires 'mobject_plugin_code' field");
  }
  if (mode === "preset" && !vizSpec.preset) errors.push("preset mode requires 'preset' field");

  // cross-check: every viz action used in an act should be implemented
  if (allActs && vizSpec.actions_implemented) {
    const implemented = new Set(vizSpec.actions_implemented);
    for (const actId of Object.keys(allActs)) {
      const act = allActs[actId];
      for (const beat of act.beats || []) {
        for (const va of beat.viz_actions || []) {
          const method = va.method;
          if (method && !implemented.has(method)) {
            errors.push(`Act ${actId} uses viz action '${method}' not in actions_implemented`);
          }
        }
      }
    }
  }

  return errors;
}


// Run every validator on a full set of artifacts. returns a map of category to
// errors, only categories with errors appear. empty map means everything is valid.
export function validateAll(
  plan: any,
  actSpecs: Record<string, any>,
  gateSpecs: Record<string, any>,
  vizSpec: any
): Record<string, string[]> {
  const results: Record<string, string[]> = {};

  const planErrors = validatePlan(plan);
  if (planErrors.length) results["plan"] = planErrors;

  for (const actId of Object.keys(actSpecs)) {
    const errs = validateActSpec(actSpecs[actId], plan);
    if (errs.length) results[`act:${actId}`] = errs;
  }

  for (const gateId of Object.keys(gateSpecs)) {
    const errs = validateGateSpec(gateSpecs[gateId], plan);
    if (errs.length) results[`gate:${gateId}`] = errs;
  }

  if (vizSpec) {
    const errs = validateVizSpec(vizSpec, plan, actSpecs);
    if (errs.length) results["viz"] = errs;
  }

  return results;
}


// CLI, mirrors validate.py's __main__: --plan, --act (repeatable), --gate (repeatable), --viz.
function main(): void {
  const argv = process.argv.slice(2);
  let planPath: string | undefined;
  const actPaths: string[] = [];
  const gatePaths: string[] = [];
  let vizPath: string | undefined;
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--plan") planPath = argv[++i];
    else if (a === "--act") actPaths.push(argv[++i]);
    else if (a === "--gate") gatePaths.push(argv[++i]);
    else if (a === "--viz") vizPath = argv[++i];
  }

  let plan: any = null;
  if (planPath) {
    plan = JSON.parse(fs.readFileSync(planPath, "utf8"));
    const errors = validatePlan(plan);
    if (errors.length) {
      console.log("PLAN errors:");
      for (const e of errors) console.log(`  - ${e}`);
    } else {
      console.log("PLAN: OK");
    }
  }

  for (const p of actPaths) {
    const spec = JSON.parse(fs.readFileSync(p, "utf8"));
    const errors = validateActSpec(spec, plan);
    const label = spec.act_id ?? p;
    if (errors.length) {
      console.log(`ACT ${label} errors:`);
      for (const e of errors) console.log(`  - ${e}`);
    } else {
      console.log(`ACT ${label}: OK`);
    }
  }

  for (const p of gatePaths) {
    const spec = JSON.parse(fs.readFileSync(p, "utf8"));
    const errors = validateGateSpec(spec, plan);
    const label = spec.gate_id ?? p;
    if (errors.length) {
      console.log(`GATE ${label} errors:`);
      for (const e of errors) console.log(`  - ${e}`);
    } else {
      console.log(`GATE ${label}: OK`);
    }
  }

  if (vizPath) {
    const spec = JSON.parse(fs.readFileSync(vizPath, "utf8"));
    const errors = validateVizSpec(spec, plan);
    if (errors.length) {
      console.log("VIZ errors:");
      for (const e of errors) console.log(`  - ${e}`);
    } else {
      console.log("VIZ: OK");
    }
  }
}

// run as a CLI only when invoked directly (not when imported).
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main();
}
